### PR TITLE
    fix #4415 feat(nimbus): support DTO v1.3.0 with app_name, app_id

### LINKED ISSUE #4415: Implement Nimbus ADR-0004 in published DTOs
[Nimbus ADR-0004](https://github.com/mozilla/nimbus-shared/blob/main/docs/adr/0004-dto-app-identifiers.md) modifies the DTO representation to deprecate the “application”  key, replacing them with "app_name," "app_id," and "channel".

For v1.3.0, we will keep `application and add:`

* `app_name` should be one of `fenix`, `firefox_desktop`
* `app_id` be one of `org.mozilla.firefox`, `org.mozilla.firefox.beta`, `org.mozilla.fenix`, or `firefox-desktop`.
* `channel` is a logical channel name, and should be one of `default`, `nightly`, `beta`, `release`

The values we choose should match the values in repositories.yaml in [https://github.com/mozilla/probe-scraper/pull/262](https://github.com/mozilla/probe-scraper/pull/262).

Corresponding bugs:

* Spec ~[https://github.com/mozilla/nimbus-shared/issues/151~](https://github.com/mozilla/nimbus-shared/issues/151~), landed as [https://github.com/mozilla/nimbus-shared/pull/152](https://github.com/mozilla/nimbus-shared/pull/152)
* Experimenter (this bug)
* Nimbus SDK SDK-169
* Jetstream

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/EXP-825)
┆Issue Number: EXP-825

### PR DESCRIPTION
**Pull request by jaredlockhart**
Because
    
* Future support for fenix experiments and glean telemetry integration requires new DTO fields for app_name and app_id
    
This commit
    
* Does away with the idea of per application channels altogether in favour of a single list of channels
* Maps channels to fenix app ids for serializing fenix experiments
* Desktop serialization behaves the same

~~Note: blocked on https://github.com/mozilla/experimenter/pull/4486~~

### CODE REVIEW

**Comment by lmorchard** [comment]
Starting to take a look

File: app/experimenter/experiments/migrations/0153_auto_20210127_1855.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("experiments", "0152_auto_20210119_1856"),
+    ]
+
+    operations = [
+        migrations.AlterField(
```
Comment (lmorchard) [inline app/experimenter/experiments/migrations/0153_auto_20210127_1855.py:13]: Since we're altering these fields and the valid choices for them, do we need a manual migration to convert any existing data in the DB?

File: app/experimenter/experiments/api/v6/serializers.py [hunk 2]
```diff
@@ -92,8 +98,16 @@ class Meta:
...
+
+    def get_app_name(self, obj):
+        return obj.application
+
+    def get_app_id(self, obj):
         if obj.is_fenix_experiment:
-            return obj.channel
+            if obj.channel in NimbusExperiment.FENIX_APP_ID:
+                return NimbusExperiment.FENIX_APP_ID[obj.channel]
+            return ""
```
Comment (k88hudson) [inline app/experimenter/experiments/api/v6/serializers.py:110]: Firefox Desktop is supposed to be `firefox-desktop` as per https://jira.mozilla.com/browse/EXP-825

**Comment by k88hudson** [comment]
Since we're looking for specific values it might be worth adding a test for each one of these configurations

```
[
  // Fenix Nightly
  {
    "application": "org.mozilla.fenix",
    "channel": "nightly",
    "appId":  "org.mozilla.fenix",
    "appName": "fenix"
  },

  // Fenix Beta
  {
    "application": "org.mozilla.firefox.beta",
    "channel": "beta",
    "appId":  "org.mozilla.firefox.beta",
    "appName": "fenix"
  },

  // Fenix Release
  {
    "application": "org.mozilla.firefox",
    "channel": "release",
    "appId":  "org.mozilla.firefox",
    "appName": "fenix"
  },

  // Desktop Nightly
  {
    "application": "firefox_desktop",
    "channel": "nightly",
    "appId":  "firefox_desktop",
    "appName": "firefox_desktop"
  },

  // Desktop Beta
  {
    "application": "firefox_desktop",
    "channel": "beta",
    "appId":  "firefox_desktop",
    "appName": "firefox_desktop"
  },

  // Desktop Release
  {
    "application": "firefox_desktop",
    "channel": "release",
    "appId":  "firefox_desktop",
    "appName": "firefox_desktop"
  },
]
```

File: app/experimenter/experiments/migrations/0153_auto_20210127_1855.py (hunk 1 above)
Comment (jaredlockhart) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py:13]: Yeah I was checking on that, staging doesn't matter, and there's only a single fenix experiment on prod with a channel so we can just change it in the admin.

File: app/experimenter/experiments/api/v6/serializers.py (hunk 2 above)
Comment (jaredlockhart) [reply app/experimenter/experiments/api/v6/serializers.py:110]: Yep it is:

https://github.com/mozilla/experimenter/pull/4487/files#diff-cc65e5b3a8e25964001f8848e456a4ffea8cfe14059fc7bf96fe133fdbd173efR42

https://github.com/mozilla/experimenter/blob/869d784733ad5dd50ca5caef7b1bd0773f2a5373/app/experimenter/experiments/constants/nimbus.py#L92

Comment (k88hudson) [reply app/experimenter/experiments/api/v6/serializers.py:110]: Ah sorry was reading this wrong

**Comment by jaredlockhart** [comment]
> Since we're looking for specific values it might be worth adding a test for each one of these configurations
> 

We coooouuuullllddddd...... Right now it's checking logical paths and treating those specifics as data, so there's two cases:

This is one case tested here:

https://github.com/mozilla/experimenter/blob/869d784733ad5dd50ca5caef7b1bd0773f2a5373/app/experimenter/experiments/tests/api/v6/test_serializers.py#L109-L125

> ```
> [
>   // Fenix Nightly
>   {
>     "application": "org.mozilla.fenix",

This is the old behaviour before this change, did we want to preserve this behaviour?

>     "channel": "nightly",
>     "appId":  "org.mozilla.fenix",
>     "appName": "fenix"
>   },
> 
>   // Fenix Beta
>   {
>     "application": "org.mozilla.firefox.beta",
>     "channel": "beta",
>     "appId":  "org.mozilla.firefox.beta",
>     "appName": "fenix"
>   },
> 
>   // Fenix Release
>   {
>     "application": "org.mozilla.firefox",
>     "channel": "release",
>     "appId":  "org.mozilla.firefox",
>     "appName": "fenix"
>   },
> 

This case is tested here:

https://github.com/mozilla/experimenter/blob/869d784733ad5dd50ca5caef7b1bd0773f2a5373/app/experimenter/experiments/tests/api/v6/test_serializers.py#L22-L94

>   // Desktop Nightly
>   {
>     "application": "firefox_desktop",
>     "channel": "nightly",
>     "appId":  "firefox_desktop",
>     "appName": "firefox_desktop"
>   },
> 
>   // Desktop Beta
>   {
>     "application": "firefox_desktop",
>     "channel": "beta",
>     "appId":  "firefox_desktop",
>     "appName": "firefox_desktop"
>   },
> 
>   // Desktop Release
>   {
>     "application": "firefox_desktop",
>     "channel": "release",
>     "appId":  "firefox_desktop",
>     "appName": "firefox_desktop"
>   },
> ]
> ```



We could add a parameterized test to test all these combinations if you'd like?

**Comment by jaredlockhart** [comment]
Oh whoops I used `app_name` and `app_id` instead of `appName` `appId`

**Comment by lmorchard** [comment]
> Since we're looking for specific values it might be worth adding a test for each one of these configurations

Is there a link to where these configurations came from? These look different than I expected.

Was reading issue #4415 and saw "The values we choose should match the values in repositories.yaml" - which I think is this: https://github.com/mozilla/probe-scraper/blob/master/repositories.yaml#L181-L517

For example, I see this in the YAML:

```yaml
  - app_name: firefox_desktop
    canonical_app_name: Firefox for Desktop
    description: The desktop version of Firefox
    url: https://github.com/mozilla/gecko-dev
    notification_emails:
      - chutten@mozilla.com
    metrics_files:
      - toolkit/components/glean/metrics.yaml
    ping_files:
      - toolkit/components/glean/pings.yaml
    dependencies:
      - org.mozilla.deprecated:glean
    channels:
      - v1_name: firefox-desktop
        app_id: firefox-desktop
```

Which makes me think we should serve something like:

```json
{
"application": "firefox_desktop",
"channel": "release",
"appId": "firefox-desktop",
"appName": "firefox_desktop"
},
```

But I'm a little confused here, because release / beta / etc aren't broken out for `firefox_desktop` like they are for `fenix`

**Comment by k88hudson** [comment]
> We could add a parameterized test to test all these combinations if you'd like?

I can see this easily being a thing we get wrong, and it's also really important to the acceptance criteria – so personally I think there would be some value in having a "serializer outputs expected app/channel info" set of of tests for each channel of each app we could add to over time

File: app/experimenter/experiments/migrations/0153_auto_20210127_1855.py (hunk 1 above)
Comment (lmorchard) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py]: This also throws an error in GQL queries:

```
    errors	[ {…} ]
    0	Object { message: "Expected a value of type \"NimbusExperimentChannel\" but received: org.mozilla.firefox.beta", path: […] }
    message	"Expected a value of type \"NimbusExperimentChannel\" but received: org.mozilla.firefox.beta"
    path	[ "experimentBySlug", "channel" ]
    0	"experimentBySlug"
    1	"channel"
```

Comment (jaredlockhart) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py]: Ooohhh iiiiiiinteresting.....  Apollo is throwing that?

Comment (jaredlockhart) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py]: Then we do have to do a data migration, good catch 👍

Comment (jaredlockhart) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py]: Yeah local just `make refresh`, prod only has a single one, but stage I didn't know it would break gql like that, so yeah we'll have to do a little db migrate.

**Comment by lmorchard** [comment]
Yeah, I think this would really benefit from a set of expected results to iterate through.

Poking around, I selected application Fenix and channel Unbranded and got this from `/api/v6/experiments/test-456/` (which seems not entirely right?):

```
{
"application":"fenix",
"app_name":"fenix",
"app_id":"",
"channel":"default"
}
```

**Comment by k88hudson** [comment]
Wow good call, app_id is indeed `firefox-desktop`.
https://github.com/mozilla/probe-scraper/blob/master/repositories.yaml  is the source of truth, I am told –

> But I'm a little confused here, because release / beta / etc aren't broken out for `firefox_desktop` like they are for `fenix`

The reason for this is that firefox desktop doesn't have different datasets per channel like mobile does

File: app/experimenter/experiments/migrations/0153_auto_20210127_1855.py (hunk 1 above)
Comment (lmorchard) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py]: > Ooohhh iiiiiiinteresting..... Apollo is throwing that?

Not Apollo, that's in the HTTP response from the django-graphene side of things

Comment (jaredlockhart) [reply app/experimenter/experiments/migrations/0153_auto_20210127_1855.py]: Oh wow!  Interesting.

**Comment by jaredlockhart** [comment]
> Yeah, I think this would really benefit from a set of expected results to iterate through.
> 
> Poking around, I selected application Fenix and channel Unbranded and got this from `/api/v6/experiments/test-456/` (which seems not entirely right?):
> 
> ```
> {
> "application":"fenix",
> "app_name":"fenix",
> "app_id":"",
> "channel":"default"
> }
> ```

No yeah this is a known edge case, there's a test for it.   @k88hudson suggested this was sufficient, it's also the same if you select 'No Channel'.  We can change it to something else if we find it's not right.

**Comment by jaredlockhart** [comment]
>   // Desktop Nightly
>   {
>     "application": "firefox_desktop",
>     "channel": "nightly",
>     "appId":  "firefox-desktop",
>     "appName": "firefox_desktop"
>   },

@k88hudson currently `application` for desktop is `firefox-desktop`, see:

https://github.com/mozilla/experimenter/blob/main/app/experimenter/experiments/constants/nimbus.py#L92

Example on prod:

https://experimenter.services.mozilla.com/api/v6/experiments/onoff-aboutwelcome-experiment/


Should this change to `firefox_desktop`?

**Comment by k88hudson** [comment]
> Should this change to firefox_desktop?

Desktop doesn't use this field, will never use this field, and it's deprecated, so go with whatever is easiest.

File: app/experimenter/experiments/tests/api/v6/test_serializers.py [hunk 3]
```diff
@@ -104,19 +107,93 @@ def test_serializers_with_feature_value_None(self):
         self.assertIsNone(serializer.data["branches"][0]["feature"]["value"])
         check_schema("experiments/NimbusExperiment", serializer.data)
 
-    def test_sets_application_channel_for_fenix_experiment(self):
+    @parameterized.expand(
+        [
```
Comment (k88hudson) [inline app/experimenter/experiments/tests/api/v6/test_serializers.py:111]: 🎉

**Review by lmorchard — APPROVED** [review]
Works functionally and the parameterized test looks like it covers the permutations, so LGTM. 

May want to see what happens with the `channel` field migration? But sounds like there's just one in production that will need tweaking?

**Comment by jaredlockhart** [comment]
oh oh yes @lmorchard the migration i forgot about that i'll add that

**Comment by k88hudson** [comment]
R+ from me on the 1.3.0 compliance

**Review by k88hudson — APPROVED** [review]
